language/move-lang/src/parser/syntax.rs

```diff
@@ -1,4 +1,1412 @@
+    let start_loc = tokens.start_loc();
+    if tokens.peek() == Tok::NameValue && tokens.lookahead()? != Tok::LBrace {
+        let v = Bind_::Var(parse_var(tokens)?);
+        let end_loc = tokens.previous_end_loc();
+        return Ok(spanned(tokens.file_name(), start_loc, end_loc, v));
+    }
+    let (ty, has_args) = parse_type_name(tokens)?;
+    let ty_args = if has_args {
+        Some(parse_type_args(tokens)?)
+    } else {
+        None
+    };
+    consume_token(tokens, Tok::LBrace)?;
+    let mut args: Vec<(Field, Bind)> = vec![];
+    while tokens.peek() != Tok::RBrace {
+        args.push(parse_bind_field(tokens)?);
+        if tokens.peek() == Tok::RBrace {
+            break;
+        }
+        consume_token(tokens, Tok::Comma)?;
+    }
+    tokens.advance()?; // consume the RBrace
+    let end_loc = tokens.previous_end_loc();
+    let unpack = Bind_::Unpack(ty, ty_args, args);
+    Ok(spanned(tokens.file_name(), start_loc, end_loc, unpack))
+}
+
+// Parse a list of bindings, which can be zero, one, or more bindings:
+//      BindList =
+//          "(" ")"
+//          | <Bind>
+//          | "(" (<Bind> ",")* <Bind> ")"
+//
+// The list is enclosed in parenthesis, except that the parenthesis are
+// optional if there is a single Bind.
+
+fn parse_bind_list<'input>(
+    tokens: &mut Lexer<'input>,
+) -> Result<BindList, ParseError<usize, &'static str>> {
+    let start_loc = tokens.start_loc();
+    let mut b: Vec<Bind> = vec![];
+    if tokens.peek() != Tok::LParen {
+        b.push(parse_bind(tokens)?);
+    } else {
+        tokens.advance()?; // consume the LParen
+        if tokens.peek() != Tok::RParen {
+            loop {
+                b.push(parse_bind(tokens)?);
+                if tokens.peek() == Tok::RParen {
+                    break;
+                }
+                consume_token(tokens, Tok::Comma)?;
+            }
+        }
+        tokens.advance()?; // consume the RParen
+    }
+    let end_loc = tokens.previous_end_loc();
+    Ok(spanned(tokens.file_name(), start_loc, end_loc, b))
+}
+
+// Parse a value:
+//      Value =
+//          <Address>
+//          | "true"
+//          | "false"
+//          | <U64>
+
+fn parse_value<'input>(
+    tokens: &mut Lexer<'input>,
+) -> Result<Value, ParseError<usize, &'static str>> {
+    let start_loc = tokens.start_loc();
+    let val = match tokens.peek() {
+        Tok::AddressValue => {
+            let addr = parse_address(tokens)?;
+            Value_::Address(addr)
+        }
+        Tok::True => {
+            tokens.advance()?;
+            Value_::Bool(true)
+        }
+        Tok::False => {
+            tokens.advance()?;
+            Value_::Bool(false)
+        }
+        Tok::U64Value => {
+            let i = u64::from_str(tokens.content()).unwrap();
+            tokens.advance()?;
+            Value_::U64(i)
+        }
+        _ => {
+            return Err(ParseError::InvalidToken {
+                location: tokens.start_loc(),
```

Comment:
I see for some of these, having the expected token is hard.... but I'm kind of hesitant to let it go in without something here. 

I want to get rid of lalrpop, but its not really blocking me to have it right now. 

Maybe we could do tests and better parser messages as the next step (i.e. soon). Thoughts?